src/storage/page_buffer.c

```diff
@@ -8017,7 +8016,23 @@ pgbuf_read_page_from_file_or_page_server (THREAD_ENTRY * thread_p, const VPID *
     }
   if (request_from_ps)
     {
-      pgbuf_request_data_page_from_page_server (vpid);
+      // *INDENT-OFF*
+      const auto replicator_lsa_ftor = [] ()
+      {
+	const passive_tran_server *const pts_ptr = get_passive_tran_server_ptr ();
+	return pts_ptr->get_replicator_lsa ();
+      };
+      // *INDENT-ON*
+
+      /* on the passive transaction server, make sure to request that the progress on the page server - from
+       * where the data page is retrieved -  be at least the same as the local progress to avoid losing
+       * any log entries being applied on both ends
+       */
+      const LOG_LSA target_repl_lsa =
+	(is_passive_transaction_server ())? replicator_lsa_ftor () : pgbuf_Pool.get_highest_evicted_lsa ();
+
+      pgbuf_request_data_page_from_page_server (vpid, target_repl_lsa);
+      // TODO: what if, by the next line, page has already been retrieved by the page broker
```

Comment:
what if?

Reply:
Page will be found immediately. See page_broker<PageT>::wait_for_page.
```suggestion
```